### ISSUE TITLE
Minor performance difference on MSR-VTT text-video retrieval

### ISSUE THREAD

**Issue by redfox94** [jayleicn/ClipBERT#26]
Hi,

Thanks for the nice work.

I am trying to replicate the experiments and observed the following as result on the text-video retrieval task with MSR-VTT dataset

```
{'text2video': {'r1': 21.0, 'r5': 45.3, 'r10': 57.2, 'medianR': 7.0, 'meanR': 59.245}, 'video2text': {'r1': 20.1, 'r5': 42.4, 'r10': 55.2, 'medianR': 8.0, 'meanR': 57.244}
```

There seems to be a minor performance difference with the reported results in 7a (with R1=22). I selected the best checkpoint on the validation set based on R1, which is at the 41500-th step (as indicated by the inference bash script below).

- Could you please kindly suggest what might be the cause? Any suggestion on how to match the reported results, please? Which checkpoint did you use for the reported results (after how many steps roughly)?
- I also notice that during the fine-tuning, retrieval performance on the validation set seems to arise and saturates rather early. Namely, it reaches around R1=10-11 very fast after a couple of thousands of steps, and remains pretty much unchanged. Is this consistent with your observation, please?

Your response is very much appreciated.

Thanks.

The finetuning scripts used:

```bash
CONFIG_PATH='src/configs/msrvtt_ret_base_resnet50.json'

horovodrun -np 4 python src/tasks/run_video_retrieval.py \
      --config $CONFIG_PATH \
      --output_dir storage
```

The config file `msrvtt_ret_base_resnet50.json` is taken from the repo without modification.
```json
  "train_datasets": [
    {
      "name": "msrvtt",
      "txt": "/txt/msrvtt_retrieval/train.jsonl",
      "img": "/img/msrvtt"
    }
  ],
  "val_datasets": [
    {
      "name": "msrvtt_retrieval",
      "txt": "/txt/msrvtt_retrieval/val.jsonl",
      "img": "/img/msrvtt"
    }
  ],
  "max_txt_len": 20,
  "max_img_size": 448,
  "img_pixel_mean": [123.675, 116.28, 103.53],
  "img_pixel_std": [1.0, 1.0, 1.0],
  "img_input_format": "BGR",
  "fps": 2,
  "num_frm": 2,
  "train_n_clips": 8,
  "score_agg_func": "lse",
  "max_n_example_per_group": 1,
  "model_config": "src/configs/base_model.json",
  "detectron2_model_cfg": "src/configs/detectron2_configs/R-50-grid.yaml",
  "e2e_weights_path": "/pretrain/clipbert_image_text_pretrained.pt",
  "detectron2_weights_path": null,
  "bert_weights_path": null,
  "tokenizer_dir": "/pretrain/bert-base-uncased/",
  "output_dir": "/storage/msrvtt_ret_default",
  "train_batch_size": 16,
  "val_batch_size": 16,
  "gradient_accumulation_steps": 1,
  "num_train_epochs": 20,
  "min_valid_steps": 100,
  "num_valid": 100,
  "learning_rate": 5e-5,
  "weight_decay": 1e-3,
  "decay": "linear",
  "optim": "adamw",
  "betas": [0.9, 0.98],
  "dropout": 0.1,
  "grad_norm": 5.0,
  "cnn_learning_rate": 5e-5,
  "cnn_weight_decay": 1e-3,
  "cnn_lr_decay": "linear",
  "seed":42,
  "fp16": 1,
  "loss_type": "ce",
  "itm_neg_size": 1
}
```

The inference command is
```bash
STEP=41500
INFERENCE_N_CLIPS=16

TXT_DB='txt_db/msrvtt_retrieval/test.jsonl'
IMG_DB='vis_db'

horovodrun -np 4 python src/tasks/run_video_retrieval.py \
      --do_inference 1 \
      --output_dir storage \
      --inference_split test \
      --inference_model_step $STEP \
      --inference_txt_db $TXT_DB \
      --inference_img_db $IMG_DB \
      --inference_batch_size 64 \
      --inference_n_clips $INFERENCE_N_CLIPS
```

**Comment by jayleicn** [comment]
Hi @redfox94, We select the best checkpoint based on mean rank, if similar, select the one with best R1. For your 2nd question, yes, we do see R1 does not change significantly after some steps, but R5/R10 are still improving.
